rest/functionsapitest/common_test.go

```diff
@@ -0,0 +1,272 @@
+
+		// Try to update using If-Match with the old etag:
+		headers := map[string]string{"If-Match": etag}
+		response = rt.SendAdminRequestWithHeaders("PUT", uri, newValue, headers)
+		assert.Equal(t, 412, response.Result().StatusCode)
+
+		// Now update successfully using the current etag:
+		headers["If-Match"] = newEtag
+		response = rt.SendAdminRequestWithHeaders("PUT", uri, newValue, headers)
+		assert.Equal(t, 200, response.Result().StatusCode)
+		newestEtag := response.HeaderMap.Get("Etag")
+		assert.Regexp(t, `"[^"]+"`, newestEtag)
+		assert.NotEqual(t, etag, newestEtag)
+		assert.NotEqual(t, newEtag, newestEtag)
+
+		// Try to delete using If-Match with the previous etag:
+		response = rt.SendAdminRequestWithHeaders("DELETE", uri, newValue, headers)
+		assert.Equal(t, 412, response.Result().StatusCode)
+
+		// Now delete successfully using the current etag:
+		headers["If-Match"] = newestEtag
+		response = rt.SendAdminRequestWithHeaders("DELETE", uri, newValue, headers)
+		assert.Equal(t, 200, response.Result().StatusCode)
+	}
+
+	t.Run("Function", func(t *testing.T) {
+		runTest(t, "/db/_config/functions/xxx", `{
+			"type": "javascript",
+			"code": "function(){return 69;}"
+		}`)
+	})
+
+	t.Run("Query", func(t *testing.T) {
+		runTest(t, "/db/_config/functions/xxxN1QL", `{
+			"type": "query",
+			"code": "select 69"
+		}`)
+	})
+
+	t.Run("Functions", func(t *testing.T) {
+		runTest(t, "/db/_config/functions", `{
+			"definitions": {
+				"zzz": {"type": "javascript", "code": "function(){return 69;}"} } }`)
+	})
+
+	t.Run("GraphQL", func(t *testing.T) {
+		runTest(t, "/db/_config/graphql", `{
+			"schema": "type Query {square(n: Int!) : Int!}", "resolvers":{}
+		}`)
+	})
+}
+
+//////// Concurrency Tests
+
+var kGraphQLTestConfig = &rest.DatabaseConfig{DbConfig: rest.DbConfig{
+	UserFunctions: &functions.FunctionsConfig{
+		Definitions: functions.FunctionsDefs{
+			"square": {
+				Type:  "javascript",
+				Code:  "function(context,args) {return args.n * args.n;}",
+				Args:  []string{"n"},
+				Allow: &functions.Allow{Channels: []string{"*"}},
+			},
+			"squareN1QL": {
+				Type:  "query",
+				Code:  "SELECT $args.n * $args.n AS square",
+				Args:  []string{"n"},
+				Allow: &functions.Allow{Channels: []string{"*"}},
+			},
+		},
+	},
+	GraphQL: &functions.GraphQLConfig{
+		Schema: base.StringPtr(`type Query { square(n: Int!): Int! }`),
+		Resolvers: map[string]functions.GraphQLResolverConfig{
+			"Query": {
+				"square": functions.FunctionConfig{
+					Type:  "javascript",
+					Code:  `function(context,args) {return args.n * args.n;}`,
+					Allow: &functions.Allow{Channels: []string{"*"}},
+				},
+			},
+		},
+	},
+}}
+
+// Asserts that running testFunc in 100 concurrent goroutines is no more than 10% slower
+// than running it 100 times in succession. A low bar indeed, but can detect some serious
+// bottlenecks, or of course deadlocks.
+func testConcurrently(t *testing.T, rt *rest.RestTester, testFunc func() bool) bool {
```

Comment:
This is a utility function that should probably go in `utilities_testing_functions_api.go`